pkg/spring/spring.go

```diff
@@ -77,7 +102,73 @@ func LoadProperties(propNames []string, configDir string, envKeyPairs []string)
 	envMap := keyPairToMap(envKeyPairs)
 	profStr := envMap["SPRING_PROFILES_ACTIVE"]
 	profs := strings.Split(profStr, ",")
-	return loadProperties(propNames, configDir, profs, envMap)
+	config, _, err := loadProperties(propNames, configDir, profs, envMap)
+	return config, err
+}
+
+// Similar to LoadDefault but provides a callback function that will be invoked when a configuration change
+// is detected. Parsing errors are also provided to the callback, so check for these as well.
+// This works by keeping track of files parsed during the initial parsing, it means that files will only
+// be tracked if they contain something. e.g. you cannot dynamically add a profile.
+// Environment variables are frozen on the initial run. This is by design
+func LoadDefaultDynamic(ctx context.Context, propNames []string, updateFn func(map[string]interface{}, error)) (map[string]interface{}, error) {
+	env := springEnv{}
+	env.initialize()
+	return loadDefaultDynamicWithEnv(env, ctx, propNames, updateFn)
+}
+
+func loadDefaultDynamicWithEnv(env springEnv, ctx context.Context, propNames []string, updateFn func(map[string]interface{}, error)) (map[string]interface{}, error) {
+	if env.configDir == "" {
+		return nil, errors.New("could not find config directory")
+	}
+
+	config, files, err := loadProperties(propNames, env.configDir, env.profiles(), env.envMap)
+	if len(files) > 0 {
+		go watchConfigFiles(ctx, files, env.envMap, updateFn)
+	}
+	return config, err
+}
+
+func watchConfigFiles(ctx context.Context, files []string, envMap map[string]string, updateFn func(map[string]interface{}, error)) {
+	watcher, err := fsnotify.NewWatcher()
+	if err != nil {
+		log.WithError(err).Errorf("unable to watch any file")
+	}
+	defer watcher.Close()
+
+	go func() {
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			case event, ok := <-watcher.Events:
+				if !ok {
+					return
+				}
+				if event.Op&fsnotify.Write == fsnotify.Write {
```

Comment:
what about linked files?

Reply:
oh i see, i guess it Symlinks get handled with fsnotify.Write 👍

Reply:
Symlink? It's taken care of by `fsnotify`. See `TestWatchSymlink`